source/common/event/dispatcher_impl.cc

```diff
@@ -157,28 +154,25 @@ DispatcherImpl::createClientConnection(Network::Address::InstanceConstSharedPtr
                                                          std::move(transport_socket), options);
 }
 
+// Create DNS resolver based on the @param typed_dns_resolver_config, which could be
+// c-ares DNS resolver, Apple DNS resolver, or any other DNS resolver type.
 Network::DnsResolverSharedPtr DispatcherImpl::createDnsResolver(
-    const std::vector<Network::Address::InstanceConstSharedPtr>& resolvers,
-    const envoy::config::core::v3::DnsResolverOptions& dns_resolver_options) {
+    const envoy::config::core::v3::TypedExtensionConfig& typed_dns_resolver_config) {
+
   ASSERT(isThreadSafe());
-#ifdef __APPLE__
-  static bool use_apple_api_for_dns_lookups =
-      Runtime::runtimeFeatureEnabled("envoy.restart_features.use_apple_api_for_dns_lookups");
-  if (use_apple_api_for_dns_lookups) {
-    RELEASE_ASSERT(
-        resolvers.empty(),
-        "defining custom resolvers is not possible when using Apple APIs for DNS resolution. "
-        "Apple's API only allows overriding DNS resolvers via system settings. Delete resolvers "
-        "config or disable the envoy.restart_features.use_apple_api_for_dns_lookups runtime "
-        "feature.");
-    RELEASE_ASSERT(!dns_resolver_options.use_tcp_for_dns_lookups(),
-                   "using TCP for DNS lookups is not possible when using Apple APIs for DNS "
-                   "resolution. Apple' API only uses UDP for DNS resolution. Use UDP or disable "
-                   "the envoy.restart_features.use_apple_api_for_dns_lookups runtime feature.");
-    return std::make_shared<Network::AppleDnsResolverImpl>(*this, api_.rootScope());
-  }
-#endif
-  return std::make_shared<Network::DnsResolverImpl>(*this, resolvers, dns_resolver_options);
+
+  // typed_dns_resolver_config need to be populated.
+  ASSERT(!typed_dns_resolver_config.name().empty() &&
```

Comment:
It's a bit hard to reason about the paths leading here; is there some guarantee that every path has already verified that user supplied configuration satisfied this?

Reply:
Thanks for the comments.

Let me remove this ASSERT() since the below function call getAndCheckFactory<Network::DnsResolverFactory>(typed_dns_resolver_config) has all the logic to check whether the name() or typed_url() is empty and throw exceptions.

Reply:
As discussed offline, I think we can get rid of `createDnsReolver()` and move this to a utility, e.g. `makeDnsResolver()`. There is no dispatcher dependency in the body of the function other than a self pointer, which can just be passed as arg. This simplifies the `Dispatcher` interface, which I think only has DNS for some historical reason. Does that work @mattklein123?

Reply:
Good idea @htuch .  
I am in the process of cleaning createDnsResolver from Dispatcher interface. Will update the PR soon.

Reply:
SG, please tag me again when this is ready for review.

Reply:
Yes, the changes with createDnsResolver() cleaned up from Dispatcher interface is updated.